In [ ]:
import numpy as np
import pandas as pd
df = pd.read_csv("stock_data.csv",parse_dates=['date'])

# Logical Statements

Booleans are the backbone of logical statements.

In [ ]:
# Using the int function
int(3.14)

In [10]:
# understanding the bool function - similar to int(object)
import numpy as np
bool()
# unless we pass 0, '', None, [], (), {} - bool function will always return True

In [ ]:
# Generating booleans
a = 5
b = 7
print(a>b)

In [ ]:
# Value of boolean
1 == True

### If Statement

In [ ]:
# if condition : 
    # Perform action (this section of code is indented)

In [ ]:
# Using booleans in if statement : Syntax if True : execute; if False : exit
if a<b : 
    print("a is less than b")

In [ ]:
# If any non-zero value is always True
if 3 :
    print("something")

In [ ]:
# Re-evaluating the NumPy .all function
my_array = np.array([1,2,3])
print((my_array>0).all()) # .all checks if all the values are True - inspiration for 'and' statement

In [ ]:
# Pandas .all and .any functions
df1 = pd.Series([1,2,3])
(df1>0)#.any()

### And Statement

Allows us to compare multiple booleans. "And" statement gets executed only if all the conditions are True.

In [ ]:
# Combining two true statements by and
True and True

In [ ]:
# Combining a true and a false boolean
True and False

In [ ]:
# More than 2 conditions
a = 1
b = 2
c = 3
if (a==1) and (b==2) and (c==3):
    print('success')

In [ ]:
# Again any value other than zero are treated as true by Python
if 3 and 4 and 5 : print('success')

### Or Statement

For the or statement to get executed, at least one of the conditions must be true.

In [ ]:
# Combining two true booleans
True or True 

In [ ]:
# Combining a true and a false boolean
True or False

In [ ]:
# More than 2 conditions
a = 1
b = 2
c = 3
if (a==1) or (b==1) or (c==1):print('success')

### If, Elif, and Else

In [ ]:
# if test1:
    # if test1 is true, do this
# elif test2:
    # if test2 is true, do this
# else : 
    # if none of the above are true, then do this

In [ ]:
# A simple example of if, elif, and else statement
a = 10

if a<10 : 
    print('A')
elif (a>=10) and (a<100) : 
    print('B')
else :
    print('C')

# Loops

Python has two loop commands - For and While loop. Operations on individual elements can be generalized using a loop in Python.

### While Loop

In [ ]:
# As long as a condition is true, While loop is executed
i=0
while i<6:
    print(i)
    i=i+1 # Can be replaced with i=i+0.5

In [ ]:
# While loops with "break"
# Break statement essentially stops running the loop if encountered
i=0
while i<5:
    # do something more meaningful
    if i==3:
        break
    print(i)
    i=i+1

In [ ]:
# While loops with "continue"
# Continue statement skips 
i=0
while i<5:
    i=i+1 # Does not matter where i=i+1 is present inside the indentation of while loop
    if i==3:continue # should be present before our mathematical operation 
    print(i)
    #print(i+i**2+i**3)

### For Loop

In [ ]:
# Defining an array and iterating through the array
my_array = np.array([1,2,3,4,5])

# Example of an opeartion 
(my_array[2]+my_array[0])**(my_array[3]-my_array[0])

In [ ]:
# Simple syntax of for loop
# Similar to if-statement, for loop starts with first indentation and ends at first un-indented line
for element in my_array:
    print(element)

In [ ]:
# range function - similar to np.arange()
range(0,5) # gives us a range object in the range class

In [ ]:
list(range(10,20,2))

In [ ]:
np.arange(10,20,2)

In [ ]:
# For loop with range
for i in range(0,len(my_array)):
    print(my_array[i])

### List Comprehension

Instead of entering data/objects inside [.] for generating lists, in list comprehensions you enter for loops and if statements within [.] to evaluate your expression. 

In [ ]:
# Printing the list we will work on
list(range(10))

In [ ]:
# Creating a new list with squares of first 5 integers
my_list1 = [] # Initiating an empty list
for i in range(10):
    my_list1.append(i**2) # Appending the square of the integer
my_list1

In [ ]:
# An alternate concise way to create lists without calling a for loop and storing output
# Syntax is [expression for items in collection if <test condition>]
my_list2 = [x**2 for x in range(10)] # saves us creating a blank list, indendations and appending
my_list2

In [ ]:
# Adding an if statement (calculating squares of even numbers only)
my_list3 = [x**2 for x in range(10) if x%2==0] 
my_list3

# Functions

Calculations and logical operations that need to be used repeatedly can be done so using a function.

In [ ]:
# Syntax for defining a function
# def function (#arguments) : return
# Similar to loops, code is grouped inside a function through indentation
def my_function():
    pass 

In [ ]:
# Functions are also objects in python
print(type(my_function))
print(id(my_function))
# Functions in python are first class objects!
# First class basically means that the function has - type, passed into other functions (as arguments), etc.

In [ ]:
# Checking the output of the function created above
my_function
#my_function()

In [ ]:
# Creating a function that does something
from math import pi
def area(x):
    """" Function returns the area of the circle """
    return 0.5*pi*x**2

In [ ]:
# Reading the docstring of the function
area?
# area(5)
# area(-5) # not good programming practice

In [ ]:
# Error when we don't pass a required variable "x"
area() 

In [ ]:
# Using the function in a list comprehension
[area(r) for r in range(10)]

In [ ]:
# Passing a numpy array inside the function
area(np.array([1,2,3]))

In [ ]:
# Modifying the area function
def area(x=10):
    """" Function returns the area of the circle """
    return 0.5*pi*x**2

In [ ]:
# Running function without specifying the key-word argument "x"
area() 

In [ ]:
# Creating a function with required and key-word arguments
def mfe_co_2022(input1,input2=5): # first required arguments and then key word arguments
    """This function returns the product of two numbers"""
    return input1*input2

In [ ]:
# Calling the function
mfe_co_2022(10.5,10)

In [ ]:
type(mfe_co_2022)

In [ ]:
# Dealing with errors we know can occur by raise  
def function(n):
    if n%1 != 0 : raise TypeError("Enter an integer!!")
    else : print(n)
    #if n<0 : raise ValueError("Your programming not so good")
    #else : print(np.sqrt(n))

function(-1)

### Lambda - Anonymous Functions

The lambda function can take many inputs, but can only have one output.

In [ ]:
# Syntax for the lambda function
# lambda inputs : output
lambda x : print(x) # A single line function only

In [ ]:
# Defining a function that squares a number
def square(x) : return x**2
square(3)

In [ ]:
# Creating a lambda function
# We are assigning the lambda function to x; So x is a function with no name
lambda i : i**2
#x = lambda i : i**2
#x(3)

In [ ]:
# Passing a numpy array through the lambda function
my_array = np.linspace(0,5,6)
my_array
fun = lambda x : np.exp(x)
fun(my_array)

### Pandas Apply

The apply function can be used on a pandas series and on a pandas dataframe. Usually, groupby, apply, and the lambda functions go together.

In [ ]:
# Creating a random dataframe
df = pd.DataFrame(np.linspace(0,10,100).reshape(20,5),columns=["A B C D E".split()])
df.head()

In [ ]:
# Apply function on a pandas data series
df[['A']].apply(np.exp) # calculating the exponential at each value of series A

In [ ]:
# Apply function on a pandas data frame
df[['A','B','C']].apply(np.max,axis=1).head() # axis=0 is along the column, axis=1 is along each row

In [ ]:
# Reloading the dataset analyzed above (shortcut to parse the dates while reading the data)
df = pd.read_csv("stock_data.csv",sep = ',',parse_dates=['date'])

# Converting any letters into NaN
df['price'] = pd.to_numeric(df['price'],errors='coerce')
df['total_returns'] = pd.to_numeric(df['total_returns'],errors='coerce')

df.head(10)

In [ ]:
# What if you want to apply a convoluted function say max-min
df[['A','B','C']].apply(np.max-np.min,axis=0)

# This is where lambda functions come handy inside the apply function
#df[['A','B','C']].apply(lambda x : x.max()-x.min(),axis=1)

In [ ]:
# Defining a function that we want to apply on the dataframe
df.groupby('permno')[['price']].apply(lambda x : x.max()/x.min())

In [ ]:
# Creating a custom function 
def modify_return(x):
    return (x.mean()-x.median())/x.std()

In [ ]:
df.groupby('permno')[['total_returns']].apply(lambda x : modify_return(x))

In [ ]:
# Apply function along the rows
#df[['price','share_outstanding']].apply(lambda x : x['price']*x['share_outstanding'],axis=1)

# Plotting

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
# Defining our two arrays 
x = np.linspace(-10,10,21)
y = x**2

In [ ]:
plt.plot(x,y)
# The x-axis is the default y-index
# First line above the plot (image location)
# Include x in the second part of plot
# Include label and legend
# Include plt.show to remove the first line 
# Change line color
# Add a Z variable
# Change line width
# Change line transparency using alpha
# Change line type by adding '--'
# Add title
# Add x-label
# Add x-ticks
# Save figure

In [ ]:
# We can also plot directly from pandas
df[df['permno']<=10029].groupby(['date','permno'])['total_returns'].mean().unstack().cumsum().plot()

### Subplots

In [ ]:
fig,ax = plt.subplots(2,2)

# Randomly defining 4 y-variables for subplots
x = np.linspace(-10,10,21)
y1 = np.exp(x)
y2 = np.sin(x)
y3 = np.power(x,3)
y4 = 10*x**2

# First subplot
ax[0,0].plot(x,y1)

# Second subplot
ax[0,1].plot(x,y2)

# Third subplot
ax[1,0].plot(x,y3)

# Fourth subplot
ax[1,1].plot(x,y4)

plt.show()

# Good Programming Practices

- Thinking through the problem before writing code is the secret. Once you have sufficiently thought through the problem, the code writes itself!
- Understanding that every program operates along multiple complexities - Time and Space complexities are the most common
- Making your code readable. Readable code that gives wrong output is better than an un-readable code that gives the right output.
- Setting your seed when you simulate random variables (very important for generating reproducible results).
- When coding up functions, make sure you give it a doc string and specify the input variable type. If there is a way for your code to break, something will eventually cause it to break.
- Commenting your code and adding doc-strings to your functions is very important when you want to revisit your code to understand what it does. It's easier to read through the comments than through code.
- Creating versions of your code (you might be interested in a Git repository). 
- Avoiding loops as much as you can and running efficient loops by vectorizing. 
- Although you feel smart when you nest multiple operations/functions on an object, long-run it is better to write one/two commands per line of code.

# Advanced Python Topics

- Git repository and Github
- Pandas - Multi index, pivot, melt, stack/unstack, pickle, concat 
- Numpy - LinAlg library, random variable distributions
- Matplotlib - 3D plots, interactive plots
- Multiprocessing (parallel computing)
- SQLite3 with Python
- Memoization
- Inheritance, Sub-classes, and meta-classes
- Map, Reduce, and Filter
- Tensors
- NumBa (easily handles extremely large data sets)
- Logging and RaiseError for Exception Handling

# Take Home Questions

- **List Comprehension**: First generate a list using the range(20) function. Next, create a second list that calculates $0.5 \cdot n \cdot (n+1)$ for each item $n$ in the first list. Solve this using a while loop and also the list comprehension method. Check if the outputs match for the two methods.

In [3]:
x=[i for i in range(20)]
[0.5*n*(n+1) for n in x]

[0.0,
 1.0,
 3.0,
 6.0,
 10.0,
 15.0,
 21.0,
 28.0,
 36.0,
 45.0,
 55.0,
 66.0,
 78.0,
 91.0,
 105.0,
 120.0,
 136.0,
 153.0,
 171.0,
 190.0]


- **Lambda function**: Create a numpy array of integers from 1 to 50. Use the lambda function to compute $x^3+x^2+x+1$ for each element $x$ of the array. Compare your output using the list comprehension method.

In [7]:
import numpy as np
np.array([i for i in range(1,51)])
fun=lambda x : x*x*x+x*x+x+1
fun(np.array([i for i in range(1,51)]))

array([     4,     15,     40,     85,    156,    259,    400,    585,
          820,   1111,   1464,   1885,   2380,   2955,   3616,   4369,
         5220,   6175,   7240,   8421,   9724,  11155,  12720,  14425,
        16276,  18279,  20440,  22765,  25260,  27931,  30784,  33825,
        37060,  40495,  44136,  47989,  52060,  56355,  60880,  65641,
        70644,  75895,  81400,  87165,  93196,  99499, 106080, 112945,
       120100, 127551])

- **For Loop**: Create a list with range(50). Return $x^3+x^2+x+1$ for every element $x$ in the list with an exception that for the last 5 elements return $x^4+x^3+x^2+x+1$. Use a for loop to do this operation. *Hint*: You might want to read up on "enumerate" and how that can be used with a for loop.

In [9]:
for index, x in enumerate([i for i in range(50)]):
    if index>=45:
        print(x*x*x*x+x*x*x+x*x+x+1)
    else:
        print(x*x*x+x*x+x+1)

1
4
15
40
85
156
259
400
585
820
1111
1464
1885
2380
2955
3616
4369
5220
6175
7240
8421
9724
11155
12720
14425
16276
18279
20440
22765
25260
27931
30784
33825
37060
40495
44136
47989
52060
56355
60880
65641
70644
75895
81400
87165
4193821
4576955
4985761
5421361
5884901


- **List comprehension with logical statements**: First generate a list using the range(30) function. Next, create a second list that creates elements based on the logic below. *Hint*: You can do this by including an if statement inside your list comprehension.

$$
f(x)=\begin{cases}
x^2    , & \text{if x is even}\\
x^3 , & \text{if x is odd}
\end{cases}
$$

In [15]:
x=[i for i in range(30)]
y=[i*i if i%2==0 else i*i*i for i in x]
y

[0,
 1,
 4,
 27,
 16,
 125,
 36,
 343,
 64,
 729,
 100,
 1331,
 144,
 2197,
 196,
 3375,
 256,
 4913,
 324,
 6859,
 400,
 9261,
 484,
 12167,
 576,
 15625,
 676,
 19683,
 784,
 24389]


- **Fibonacci sequence**: Construct a function "fibonacci" that takes in the required variable integer "$n$" and returns the $n_{th}$ term in a Fibonacci sequence (1,1,2,3,5,...). For example, if your call your function and pass in the argument n = 6, i.e fibonacci(n=6) it should return the value 8. You can use a for or a while loop inside your function. Print the function output for n=30, 50, and 100. 

In [17]:
def f(n):
    a,b=0,1
    while n>1:
        a,b=b,a+b
        n-=1
    return b
print(f(6))
print(f(30))
print(f(50))
print(f(100))

8
832040
12586269025
354224848179261915075


- **Net Present Value**: Code up a function that computes the net present value of a stream of cash-flows and call this function NPV. The function will take three inputs - cash-flows, dates, and interest rate (constant). We will use the formula specified below to compute the NPV of a stream of cash-flows on specific dates. The time periods $t_1,t_2,..,t_N$ are computed as differences from first date. Make sure you annualize the difference in days (convert them to years). Check out this wikipedia link if you want to read up some more (https://en.wikipedia.org/wiki/Net_present_value). 

$$NPV(CF) = CF_0 + \frac{CF_1}{(1+r)^{t_1}} + \frac{CF_2}{(1+r)^{t_2}} + ... \frac{CF_N}{(1+r)^{t_N}}$$

Run this function for the following input 

$$\text{Compute : } \text{NPV}(\text{CF} = \text{[-100,50,40,30]},\text{date} = [\text{01-Jan-2020,31-Mar-2021,30-Jun-2021,31-Dec-2024}],r=0.05)$$

In [23]:
from datetime import datetime

def npv(cd,date,r):
    time1=datetime.strptime(date[0],'%d-%b-%Y')
    time2=datetime.strptime(date[1],'%d-%b-%Y')
    time3=datetime.strptime(date[2],'%d-%b-%Y')
    time4=datetime.strptime(date[3],'%d-%b-%Y')
    t1=(time2-time1).days/365
    t2=(time3-time2).days/365
    t3=(time4-time3).days/365
    return cd[0]+cd[1]/((1+r)**t1)+cd[2]/((1+r)**t2)+cd[3]/((1+r)**t3)


npv([-100,50,40,30],['01-Jan-2020','31-Mar-2021','30-Jun-2021','31-Dec-2024'],0.05) 


50 1.062708263812578


11.848108716434613

- **Pandas .rolling()** is a very useful function to know (check out the doc string of pandas.Series.rolling for a better understanding of what the function does). In this problem, we will calculate the 1-year rolling sum of returns for the data in *stock_data.csv* file. Clean the data (remove strings and drop NaN values). Use the pandas rolling function (window=12) to calculate the sum of 1-year returns. Add this column to the original dataframe. 

In [42]:
import pandas as pd
df=pd.read_csv("stock_data.csv")
df.dropna().reset_index(drop=True,inplace=True)
df['total_returns']=pd.to_numeric(df['total_returns'],errors='coerce')
df['date']=pd.to_datetime(df['date'])
roll=df.groupby(['permno'])['total_returns'].rolling(12).sum().reset_index()
df.merge(roll, on )

permno  level_1  total_returns
0      10001        0            NaN
1      10001        1            NaN
2      10001        2            NaN
3      10001        3            NaN
4      10001        4            NaN
5      10001        5            NaN
6      10001        6            NaN
7      10001        7            NaN
8      10001        8            NaN
9      10001        9            NaN
10     10001       10            NaN
11     10001       11       0.212478
12     10001       12       0.102148
13     10001       13       0.150072
14     10001       14       0.173455
15     10001       15       0.225216
16     10001       16       0.062042
17     10001       17       0.085263
18     10001       18       0.136945
19     10001       19       0.025473
20     10001       20       0.122872
21     10001       21       0.051321
22     10001       22       0.023511
23     10001       23       0.136947
24     10001       24       0.136491
25     10001       25       0.087797
26     10001       26       0.070075
27     10001       27       0.046431
28     10001       28       0.070220
29     10001       29      -0.042191
...      ...      ...            ...
2930   10137     2930      -0.284544
2931   10137     2931      -0.625317
2932   10137     2932      -0.460119
2933   10137     2933      -0.542374
2934   10137     2934      -0.421713
2935   10137     2935      -0.635663
2936   10137     2936      -0.648804
2937   10137     2937      -0.595463
2938   10137     2938      -0.648615
2939   10137     2939      -0.534606
2940   10137     2940      -0.517635
2941   10137     2941      -0.406605
2942   10137     2942      -0.211232
2943   10137     2943      -0.170711
2944   10137     2944      -0.376675
2945   10137     2945      -0.266430
2946   10137     2946      -0.355870
2947   10137     2947       0.014085
2948   10137     2948       0.049696
2949   10137     2949      -0.122035
2950   10137     2950      -0.147147
2951   10137     2951      -0.161063
2952   10137     2952      -0.041395
2953   10137     2953      -0.099960
2954   10137     2954      -0.015792
2955   10137     2955       0.069891
2956   10137     2956            NaN
2957   10137     2957            NaN
2958   10137     2958            NaN
2959   10137     2959            NaN

[2960 rows x 3 columns]




- **Combining Pandas apply and lambda functions**: Load and clean the stock_data.csv file. Compute the market cap by multiplying the price and outstanding shares. You need to segregate PERMNOs into large-cap, mid-cap, and small-cap stocks. First, you have to write a function that takes in a value $x$ (market cap) as an input. For segregating use the following criteria mentioned below. Use the pandas apply along with lambda function, to pass the function pass you have just coded up.

$$
f(x)=\begin{cases}
\text{small cap}    , & \text{if x} < 100000\\
\text{mid cap}    , & 100000 \le \text{if x} < 1000000\\
\text{large cap} , & \text{if x} \ge  1000000
\end{cases}
$$

- **Plot and Subplots**: For the 4 PERMNOs - (10137, 10051, 10057, 10028), calculate the cumulative sum of returns and plot them in a 4-by-4 sub plot. Also, plot all their cumulative sum of returns in one plot. *Hint*: To compute the cumulative sum, you can use the .cumsum() function of pandas. You can also use the .plot() function of pandas dataframe instead of using matplotlib's plt function.

- **(optional) Memoization**: A simple way to create a fibonacci sequence without using loops is by coding it as shown in the code chunk below. Carefully understand how this function does the same job as your function that you coded above. 

In [ ]:
def fibonacci(n):
    """Function takes input integer 'n' and returns the n-th sequence of the fibonacci series"""
    if n<=2 : return(1)
    else : return fibonacci(n-1)+fibonacci(n-2)

First, create a range of values of $n$ from 1 to 50 and loop this function over the possible integer values of $n$. What do you observe? Why does the loop start laboring after n=30? 

Second, we will start storing our computations of fibonacci function in a dictionary and use the function for computing previously un-computed values of the function. We can do this by using the new function "fibonacci_new" coded below - 

In [17]:
# Task 3 - creating a Fibonacci sequence without using loops (using Memoization)
fibonacci_store = {}

def fibonacci_new(n):
    """Function takes input integer 'n' and returns the n-th sequence of the fibonacci series (with memoization)"""
    
    # Checking if the value n is in our cache
    if n in fibonacci_store : return fibonacci_store[n]
    
    # If not, then calculate the n-th value
    if n <= 2 : return 1
    else : 
        value = fibonacci_new(n-1)+fibonacci_new(n-2)
        fibonacci_store[n] = value # Storing the newly computed value of the function in our dictionary
        return value

Again create a range of values for $n$ from 1 to 50 and loop this new function over the possible integer values of $n$. What do you observe now? Why is this function not struggling unlike the "fibonacci()" function? 

Our actions of creating a dictionary and storing the values is coded up for us in the form of an inbuilt function in Python. This will involve importing the function "lru_cache" from the library "functools" as shown below. You can test this function below like you have tested the first fibonacci function. What do you observe?

In [ ]:
from functools import lru_cache

@lru_cache(maxsize=1000)
def fibonacci(n):
    """Function takes input integer 'n' and returns the n-th sequence of the fibonacci series (no loops & no memoization)"""
    if n <= 2 : return 1
    else : return fibonacci(n-1)+fibonacci(n-2)